broker_sync_test.go

```diff
@@ -0,0 +1,361 @@
+	dstStore := dssync.MutexWrap(datastore.NewMapDatastore())
+	srcHost := mkTestHost()
+	srcLnkS := mkLinkSystem(srcStore)
+	testLnkS := mkLinkSystem(testStore)
+	lp, err := NewPublisher(context.Background(), srcHost, srcStore, srcLnkS, testTopic)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	chainLnks := mkChain(testLnkS, true)
+
+	dstHost := mkTestHost()
+	srcHost.Peerstore().AddAddrs(dstHost.ID(), dstHost.Addrs(), time.Hour)
+	dstHost.Peerstore().AddAddrs(srcHost.ID(), srcHost.Addrs(), time.Hour)
+	dstLnkS := mkLinkSystem(dstStore)
+
+	lb, err := NewLegBroker(dstHost, dstStore, dstLnkS, testTopic, nil, 0, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err = lb.SetLatestSync(srcHost.ID(), chainLnks[3].(cidlink.Link).Cid)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if err := srcHost.Connect(context.Background(), dstHost.Peerstore().PeerInfo(dstHost.ID())); err != nil {
+		t.Fatal(err)
+	}
+
+	mkChain(srcLnkS, true)
+
+	time.Sleep(1 * time.Second)
+
+	watcher, cncl := lb.OnChange()
+
+	t.Cleanup(func() {
+		lp.Close()
+		lb.Close()
+		cncl()
+	})
+
+	// Fetching first few nodes.
+	newBrokerUpdateTest(t, lp, lb, dstStore, watcher, srcHost.ID(), chainLnks[2], false, chainLnks[2].(cidlink.Link).Cid)
+
+	// Check that first nodes hadn't been synced
+	if _, err := dstStore.Get(datastore.NewKey(chainLnks[3].(cidlink.Link).Cid.String())); err != datastore.ErrNotFound {
+		t.Fatalf("data should not be in receiver store: %v", err)
+	}
+
+	// Set latest sync so we pass through one of the links
+	err = lb.SetLatestSync(srcHost.ID(), chainLnks[1].(cidlink.Link).Cid)
+	if err != nil {
+		t.Fatal(err)
+	}
+	assertBrokerLatestSyncEquals(t, lb, srcHost.ID(), chainLnks[1].(cidlink.Link).Cid)
+	// Update all the chain from scratch again.
+	newBrokerUpdateTest(t, lp, lb, dstStore, watcher, srcHost.ID(), chainLnks[0], false, chainLnks[0].(cidlink.Link).Cid)
+
+	// Check if the node we pass through was retrieved
+	if _, err := dstStore.Get(datastore.NewKey(chainLnks[1].(cidlink.Link).Cid.String())); err != datastore.ErrNotFound {
+		t.Fatalf("data should not be in receiver store: %v", err)
+	}
+}
+func TestBrokerStepByStepSync(t *testing.T) {
+	srcStore := dssync.MutexWrap(datastore.NewMapDatastore())
+	dstStore := dssync.MutexWrap(datastore.NewMapDatastore())
+	srcHost := mkTestHost()
+	srcLnkS := mkLinkSystem(srcStore)
+	lp, err := NewPublisher(context.Background(), srcHost, srcStore, srcLnkS, testTopic)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	dstHost := mkTestHost()
+	srcHost.Peerstore().AddAddrs(dstHost.ID(), dstHost.Addrs(), time.Hour)
+	dstHost.Peerstore().AddAddrs(srcHost.ID(), srcHost.Addrs(), time.Hour)
+	dstLnkS := mkLinkSystem(dstStore)
+
+	lb, err := NewLegBroker(dstHost, dstStore, dstLnkS, testTopic, nil, 0, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if err := srcHost.Connect(context.Background(), dstHost.Peerstore().PeerInfo(dstHost.ID())); err != nil {
+		t.Fatal(err)
+	}
+
+	time.Sleep(1 * time.Second)
```

Comment:
Yes, that has been something that has been requested for a long time so that we can avoid guessing at sleep times (really, sleeps should never be in tests if it can be avoided).  I think Daniel may have put a fix in a later version of libp2p-pubsub, but we need to see if we can upgrade to that version yet.  Another PR.

Reply:
Replaced waiting for mesh using time.Sleep() with a function that can be updated when a better way is available.